backend: use MIR for representing constant expressions

zerbina opened the pull request:
## Summary

Instead of `PNode`, use a subset of the MIR for representing constant
expression in the mid- and back-end. Apart from removing a significant
amount of `PNode` usage from the backend, using a dedicated IR for
constant expression paves the way for applying transformation and
lowering passes to them.

## Details

### Constant expression IR

* a subset of the MIR is used (refer to the `Constant Expressions`
  section in `doc/mir.rst`)
* in the subset, recursion is allowed
* using a subset of the MIR allows for re-using most of the existing
  MIR facilities (changesets, traversal routines, etc.)

### Translation

* the translation is implemented as a single-pass traversal over the
  constant expression AST
* since procedures referenced by constant expression are now registered
  with the `MirEnv` as they're encountered, the separate scanning pass
  (`scanExpr`) is obsolete
* the translation also handles `nkHiddenStdConv` and `nkHiddenSubConv`
  nodes, making the custom `mirgen.isDeepConstExpr` overload, which
  treated these nodes as non-constant, obsolete
* the constant-expression-to-`CgNode` translation (`compat.translate`)
  is updated to operate on the MIR constant expression

### VM serialization

The VM itself is not affected, but the logic for initializing memory
locations directly from constant expressions is. Initializing a VM
memory location directly from a MIR constant expression is implemented
by the new `vmserialize` module.

Apart from the loading of constant data in `vmjit.updateEnvironment`
(which must use `vmserialize`, since it now gets passed a `MirTree`),
there were two other situations where a VM memory location is
initialized from a `PNode` expression:
1. for `static` macro parameter initialization (`putIntoReg`)
2. for setting the value of a global via the compiler API
  (`setGlobalValue`)

Both are changed to first translate the `PNode` AST to a MIR constant
expression, which the register or memory location is then initialized
with. This has the benefits that:
- initializing VM locations from constant expression works the same
  everywhere
- procedures referenced from the AST are automatically registered with
  the `MirEnv` and JIT environment
- the `vmcompilerserdes` procedures for initializing a VM location from
  `PNode` AST become obsolete

Since they're now obsolete, the `vmcompilerserdes` serialization
routines are all removed. This has an important implication: a
`LinkerData` instance doesn't have to be carried around by `vmdef.TCtx`
anymore. Nothing is removed here yet -- the cleanup is left to follow-
up work.

Finally, the serialization of constant expressions in `packed_env` is
also updated to operate directly on MIR constant expressions.

### Misc

* speed up `mirconstr.pop` in the full-buffer-to-empty-buffer case
  (which happens for `constDataToMir`)

zerbina committed:
datatables: the table now stores `MirTree`s

The interface and most of the implementation stays the same, but
instead of `PNode`, the new MIR constant expression trees are used.

zerbina committed:
mir: specify how MIR constant expression look like

zerbina committed:
mirgen: implement constant-to-MIR translation

The new `constDataToMir` procedure translates the AST of a constant
expression to its corresponding MIR constant expression.

zerbina committed:
mirconstr: more efficient popping in edge case

The nodes don't need to be moved in case of a full-buffer-to-empty-
buffer move; the buffers can simply be swapped instead.

zerbina committed:
backends: update constant data translation

The AST cannot be placed directly into the `DataTable` -- it has to be
translated first, which is done via `mirgen.constDataToMir`.

zerbina committed:
compat: change `translate` to process `MirTree`s

`translate` is the procedure responsible for translating constant
expression to the `CgNode` IR. Since constant expressions now use the
MIR, so has `translate`.

zerbina committed:
vmjit: use `initFromExpr`

The `PNode`-based serialization routines from `vmcompilerserdes`
cannot be used anymore, now that constant expressions use the MIR.

zerbina committed:
implement initializing VM memory from MIR constants

The new `vmserialize` modules provides routines for directly
initializing a VM memory location with MIR constant expressions,
which is what will be used instead of `vmcompilerserdes.serialize`.

zerbina committed:
vmcompilerserdes: remove `serialize`

The `PNode`-based initialization logic is completely obsolete now.
Everything related to it is removed.

zerbina committed:
compilerbridge: use `initFromExpr` for `setGlobalValue`

This removes the last user of the `PNode`-based initialization.

zerbina committed:
compilerbridge: use `initFromExpr` for `putIntoReg`

Instead of using the `PNode`-based initialization, the expression is
first translated to the MIR, and then loaded into the location with
`initFromExpr`.

The new `constDataToMir` procedure provided by `vmjit` delegates to
the normal AST -> MIR node translation, while also taking care of the
registering procedures discovered during the translation with the VM
environment.

This `vmaux.lookupProc` procedure is obsolete and thus removed.

zerbina committed:
mirgen: remove the custom `isDeepConstExpr` variant

The AST -> MIR translation for constant data handles conversion
expression, meaning that the `isDeepConstExpr` workaround is
obsolete.

zerbina committed:
packed_env: data serialization works with `MirTree`

Instead of from `PNode`, the serialization logic for constant data now
operates directly on MIR constant expressions.

zerbina committed:
ccgexprs: allow expanded nil closures

These constructions are valid, and they have to work. Since constant
expression involving closure construction only came from
`vmcompilerserdes`, the assertion was not a problem, as all 'nil'
closure are turned into an `nkNilLit` there.

Clyybber commented on compiler/mir/datatables.nim:
```diff
@@ -32,40 +32,78 @@ type
+    return false
+
+  # both trees are known to have the same length at this point
+  var i = 0
+  # traverse both trees simultaneously and look for a divergence:
+  while i < a.len:
+    if a[i] != b[i]:
+      return false # divergence -> not equal
+
+    inc i
```
could use a for loop here
```suggestion
  # traverse both trees simultaneously and look for a divergence:
  for i in 0..<a.len:
    if a[i] != b[i]:
      return false # divergence -> not equal
```

zerbina replied:
Oh, nice catch.

zerbina committed:
datatables: replace `while` loop with `for` loop

Co-authored-by: Clyybber <darkmine956@gmail.com>

zerbina committed:
Merge branch 'devel' into mir-constants

zerbina committed:
mirpasses: update `extractStringLiterals`

Passing a `PNode` to `datatables.getOrPut` no longer works -- it has to
be a `MirTree`.

saem commented:
/merge
